ouroboros-network/src/Ouroboros/Network/PeerSelection/Churn.hs

```diff
@@ -61,180 +77,369 @@ peerChurnGovernor tracer deadlineChurnInterval bulkChurnInterval psOverallTimeou
+    decreaseActiveBigLedgerPeers mode targets =
+      targets {
+        targetNumberOfActiveBigLedgerPeers =
+          case mode of
+            ChurnModeNormal ->
+              decrease $ targetNumberOfActiveBigLedgerPeers base
+            ChurnModeBulkSync ->
+              min 1 (targetNumberOfActiveBigLedgerPeers base)
+      }
+
+    checkActiveBigLedgerPeersDecreased
+      :: CheckPeerSelectionCounters
+    checkActiveBigLedgerPeersDecreased
+      PeerSelectionCounters { numberOfActiveBigLedgerPeers }
+      PeerSelectionTargets { targetNumberOfActiveBigLedgerPeers }
+      =
+      numberOfActiveBigLedgerPeers <= targetNumberOfActiveBigLedgerPeers
+
+
+    decreaseEstablishedBigLedgerPeers
+      :: ModifyPeerSelectionTargets
+    decreaseEstablishedBigLedgerPeers targets =
+      targets {
+        targetNumberOfEstablishedBigLedgerPeers =
+          decrease (targetNumberOfEstablishedBigLedgerPeers base -
+                    targetNumberOfActiveBigLedgerPeers base)
+          + targetNumberOfActiveBigLedgerPeers base
+      }
+
+    checkEstablishedBigLedgerPeersDecreased
+      :: CheckPeerSelectionCounters
+    checkEstablishedBigLedgerPeersDecreased
+      PeerSelectionCounters { numberOfEstablishedBigLedgerPeers }
+      PeerSelectionTargets { targetNumberOfEstablishedBigLedgerPeers }
+      =
+      numberOfEstablishedBigLedgerPeers <= targetNumberOfEstablishedBigLedgerPeers
+
+
+    decreaseKnownPeers
+      :: ModifyPeerSelectionTargets
+    decreaseKnownPeers targets =
+      targets {
+          targetNumberOfRootPeers =
+            decrease (targetNumberOfRootPeers base - targetNumberOfEstablishedPeers base)
+            + targetNumberOfEstablishedPeers base
+        , targetNumberOfKnownPeers =
+            decrease (targetNumberOfKnownPeers base - targetNumberOfEstablishedPeers base)
+            + targetNumberOfEstablishedPeers base
+        , targetNumberOfKnownBigLedgerPeers =
+            decrease (targetNumberOfKnownBigLedgerPeers base -
+                      targetNumberOfEstablishedBigLedgerPeers base)
+            + targetNumberOfEstablishedBigLedgerPeers base
+        }
+
+    checkKnownPeersDecreased
+      :: PeerSelectionCounters -> PeerSelectionTargets -> Bool
+    checkKnownPeersDecreased
+      PeerSelectionCounters { numberOfKnownPeers,
+                              numberOfKnownBigLedgerPeers }
+      PeerSelectionTargets { targetNumberOfKnownPeers,
+                             targetNumberOfKnownBigLedgerPeers }
+      =
+        -- note: we are not checking target root peers, since it is a one-sided
+        -- target
+         numberOfKnownPeers <= targetNumberOfKnownPeers
+      && numberOfKnownBigLedgerPeers <= targetNumberOfKnownBigLedgerPeers
+
+
+    increaseKnownPeers
+      :: ModifyPeerSelectionTargets
+    increaseKnownPeers targets =
+      targets {
+          targetNumberOfRootPeers =
+            decrease (targetNumberOfRootPeers base - targetNumberOfEstablishedPeers base)
```

Comment:
oops ;) I think this should be targetNumberOfRootPeers = targetNumberOfRootPeers base
and analogously below:
targetNumberOfKnownPeers = targetNumberOfKnownPeers base
targetNumberOfKnownBigLedgerPeers = targetNumberOfKnownBigLedgerPeers base